selene_sdk/samplers/random_positions_sampler.py

```diff
@@ -151,86 +159,214 @@ def __init__(self,
             save_datasets=save_datasets,
             output_dir=output_dir)
 
-        self._sample_from_mode = {}
-        self._randcache = {}
-        for mode in self.modes:
-            self._sample_from_mode[mode] = None
-            self._randcache[mode] = {"cache_indices": None, "sample_next": 0}
-
         self.sample_from_intervals = []
         self.interval_lengths = []
 
-        if self._holdout_type == "chromosome":
-            self._partition_genome_by_chromosome()
+        self._num_chroms = 0
+        self._genome_n_bases = 0
+        # Holds the size of the parition of each mode (the portion that is
+        # specified to belong to each mode)
+        self._parititon_sizes = {"validate" : 0, "train" : 0, "test" : 0}
+
+        self._validation_holdout = validation_holdout
+        self._N_validation = validation_size
+
+        if test_holdout:
+            self._test_holdout = test_holdout
+            self._N_test = test_size
+
+        self._N_train = train_size
+
+        # Holds numpy array data sets for each mode (train, validate, test)
+        # which is made upon initilization of the class.
+        # These samples' initialize size is determined by the sample size
+        # designated by the class instance and is subsequently shrunken as
+        # calls to sample are made.
+        self._partition_ixs = {"train": np.zeros(self._N_train, dtype=np.int64),
+                                "validate": np.zeros(self._N_validation, dtype=np.int64),
+                                "test": np.zeros(self._N_test, dtype=np.int64)}
+
+        # Count the number of chromosomes and bases. If sampling by chromosome,
+        # count the number of base pairs in the chromosomes per mode.
+        for chrom, len_chrom in self.reference_sequence.get_chr_lens():
+            if not isinstance(validation_holdout, float):
+                if chrom in self.validation_holdout:
+                    self._parititon_sizes["validate"] += len_chrom
+                elif chrom in self.test_holdout:
+                    self._parititon_sizes["test"] += len_chrom
+                else:
+                    self._parititon_sizes["train"] += len_chrom
+            self._num_chroms += 1
+            self._genome_n_bases += len_chrom
```

Comment:
also, don't initialize `_num_chroms` and `_genome_n_bases` so early if you don't plan to do anything with them until this for-loop. might as well put the variables close to where they are updated